tuf/tuf.go

```diff
@@ -294,114 +272,104 @@ func (tr *Repo) GetAllLoadedRoles() []*data.Role {
 		return err
 	}
 
 	// check the parent role's metadata
-	p, ok := tr.Targets[parent]
+	_, ok := tr.Targets[parent]
 	if !ok { // the parent targetfile may not exist yet - if not, then create it
 		var err error
-		p, err = tr.InitTargets(parent)
+		_, err = tr.InitTargets(parent)
 		if err != nil {
 			return err
 		}
 	}
 
-	for _, k := range keys {
-		if !utils.StrSliceContains(role.KeyIDs, k.ID()) {
-			role.KeyIDs = append(role.KeyIDs, k.ID())
+	// Walk to parent, and either set or update this delegation
+	// Ensure all updates are valid
+	delegationUpdateVisitor := func(tgt *data.SignedTargets, validRole data.DelegationRole) error {
+		var err error
+		var createdRole bool
+		// Validate the changes underneath this restricted validRole for adding paths, reject invalid path additions
+		if len(changes.AddPaths) != len(data.RestrictDelegationPathPrefixes(validRole.Paths, changes.AddPaths)) {
+			return data.ErrInvalidRole{Role: roleName, Reason: "invalid paths to add to role"}
+		}
+		// Try to find the delegation and amend it using our changelist
+		var delgRole *data.Role
+		for _, role := range tgt.Signed.Delegations.Roles {
+			if role.Name == roleName {
+				delgRole = role
+				delgRole.AddPaths(changes.AddPaths)
+				delgRole.RemovePaths(changes.RemovePaths)
+				if changes.ClearAllPaths {
+					delgRole.Paths = []string{}
+				}
+				delgRole.RemoveKeys(changes.RemoveKeys)
+			}
 		}
-		p.Signed.Delegations.Keys[k.ID()] = k
-	}
-
-	// if the role has fewer keys than the threshold, it
-	// will never be able to create a valid targets file
-	// and should be considered invalid.
-	if len(role.KeyIDs) < role.Threshold {
-		return data.ErrInvalidRole{Role: role.Name, Reason: "insufficient keys to meet threshold"}
+		// We didn't find the role earlier, so create it
+		if delgRole == nil {
+			delgRole, err = changes.ToNewRole(roleName)
+			if err != nil {
+				return err
+			}
+			createdRole = true
+		}
+		// Add the key IDs to the role and the keys themselves to the parent
+		for _, k := range changes.AddKeys {
+			if !utils.StrSliceContains(delgRole.KeyIDs, k.ID()) {
+				delgRole.KeyIDs = append(delgRole.KeyIDs, k.ID())
+			}
+			tgt.Signed.Delegations.Keys[k.ID()] = k
+		}
+		// Make sure we have a valid role still
+		if len(delgRole.KeyIDs) < delgRole.Threshold {
+			return data.ErrInvalidRole{Role: roleName, Reason: "insufficient keys to meet threshold"}
+		}
+		if createdRole {
+			tgt.Signed.Delegations.Roles = append(tgt.Signed.Delegations.Roles, delgRole)
+		}
+		tgt.Dirty = true
+		utils.RemoveUnusedKeys(tgt)
+		return ErrStopWalk{}
 	}
 
-	foundAt := utils.FindRoleIndex(p.Signed.Delegations.Roles, role.Name)
-
-	if foundAt >= 0 {
-		p.Signed.Delegations.Roles[foundAt] = role
-	} else {
-		p.Signed.Delegations.Roles = append(p.Signed.Delegations.Roles, role)
+	// Walk to the parent of this delegation, since that is where its role metadata exists
+	err := tr.WalkTargets("", parent, delegationUpdateVisitor)
```

Comment:
Add comment: we don't have to verify that the walker reached its desired role in this scenario since we've already done a walk earlier to the parent role